verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "verilog/tools/ls/symbol-table-handler.h"
+
+#include <filesystem>
+
+#include "absl/flags/flag.h"
+#include "common/strings/line_column_map.h"
+#include "common/util/file_util.h"
+#include "verilog/analysis/verilog_filelist.h"
+
+ABSL_FLAG(std::string, file_list_path, "verible.filelist",
+          "Name of the file with Verible FileList for the project");
+
+namespace verilog {
+
+static constexpr absl::string_view kFileSchemePrefix = "file://";
+
+absl::string_view LSPUriToPath(absl::string_view uri) {
+  if (!absl::StartsWith(uri, kFileSchemePrefix)) return "";
+  return uri.substr(kFileSchemePrefix.size());
+}
+
+std::string PathToLSPUri(absl::string_view path) {
+  std::filesystem::path p(path.begin(), path.end());
+  return absl::StrCat(kFileSchemePrefix, std::filesystem::absolute(p).string());
+}
+
+void SymbolTableHandler::SetProject(
+    const std::shared_ptr<VerilogProject> &project) {
+  currproject = project;
+  ResetSymbolTable();
+  LoadProjectFileList(currproject->TranslationUnitRoot());
+}
+
+void SymbolTableHandler::ResetSymbolTable() {
+  checkedfiles.clear();
+  symboltable = std::make_unique<SymbolTable>(currproject.get());
+}
+
+void SymbolTableHandler::BuildSymbolTableFor(const VerilogSourceFile &file) {
+  auto result = BuildSymbolTable(file, symboltable.get(), currproject.get());
+}
+
+void SymbolTableHandler::BuildProjectSymbolTable() {
+  ResetSymbolTable();
+  if (!currproject) {
+    return;
```

Comment:
What if we do not _have_ a `currproject` set and it is `nullptr` ?
We should probably return a boolean here, so that we can use that to skip anything afterwards.